syz-manager/rpc.go

```diff
@@ -262,161 +173,156 @@ func (serv *RPCServer) Check(a *rpctype.CheckArgs, r *int) error {
+	r.DropMaxSignal = runner.instModules.Decanonicalize(dropSignal.ToRaw())
 
-	serv.mu.Lock()
-	defer serv.mu.Unlock()
+	log.Logf(1, "exchange with %s: %d done, %d new requests, %d new max signal, %d drop signal",
+		a.Name, len(a.Results), len(r.Requests), len(r.NewMaxSignal), len(r.DropMaxSignal))
+	return nil
+}
 
-	f := serv.fuzzers[a.Name]
-	if f == nil {
-		// This is possible if we called shutdownInstance,
-		// but already have a pending request from this instance in-flight.
-		log.Logf(1, "poll: fuzzer %v is not connected", a.Name)
+func (serv *RPCServer) updateFilteredCover(pcs []uint32) error {
+	if len(pcs) == 0 || serv.coverFilter == nil {
 		return nil
 	}
-	newMaxSignal := serv.maxSignal.Diff(a.MaxSignal.Deserialize())
-	if !newMaxSignal.Empty() {
-		serv.maxSignal.Merge(newMaxSignal)
-		serv.stats.maxSignal.set(len(serv.maxSignal))
-		for _, f1 := range serv.fuzzers {
-			if f1 == f || f1.rotated {
-				continue
-			}
-			f1.newMaxSignal.Merge(newMaxSignal)
+	// Note: ReportGenerator is already initialized if coverFilter is enabled.
+	rg, err := getReportGenerator(serv.cfg, serv.modules)
+	if err != nil {
+		return err
+	}
+	filtered := 0
+	for _, pc := range pcs {
+		if serv.coverFilter[uint32(rg.RestorePC(pc))] != 0 {
+			filtered++
 		}
 	}
-	if f.rotated {
-		// Let rotated VMs run in isolation, don't send them anything.
+	serv.stats.corpusCoverFiltered.add(filtered)
+	return nil
+}
+
+func (serv *RPCServer) shutdownInstance(name string) []byte {
+	var runner *Runner
+	if val, _ := serv.runners.LoadAndDelete(name); val != nil {
+		runner = val.(*Runner)
+	} else {
 		return nil
 	}
-	r.MaxSignal = f.newMaxSignal.Split(2000).Serialize()
-	if a.NeedCandidates {
-		r.Candidates = serv.mgr.candidateBatch(serv.batchSize)
-	}
-	if len(r.Candidates) == 0 {
-		batchSize := serv.batchSize
-		// When the fuzzer starts, it pumps the whole corpus.
-		// If we do it using the final batchSize, it can be very slow
-		// (batch of size 6 can take more than 10 mins for 50K corpus and slow kernel).
-		// So use a larger batch initially (we use no stats as approximation of initial pump).
-		const initialBatch = 50
-		if len(a.Stats) == 0 && batchSize < initialBatch {
-			batchSize = initialBatch
-		}
-		for i := 0; i < batchSize && len(f.inputs) > 0; i++ {
-			last := len(f.inputs) - 1
-			r.NewInputs = append(r.NewInputs, f.inputs[last])
-			f.inputs[last] = rpctype.Input{}
-			f.inputs = f.inputs[:last]
-		}
-		if len(f.inputs) == 0 {
-			f.inputs = nil
-		}
+
+	runner.mu.Lock()
+	if runner.requests == nil {
+		// We are supposed to invoke this code only once.
+		panic("Runner.requests is already nil")
 	}
-	for _, inp := range r.NewInputs {
-		inp.Cover, inp.Signal = f.instModules.Decanonicalize(inp.Cover, inp.Signal)
+	oldRequests := runner.requests
+	runner.requests = nil
+	runner.mu.Unlock()
+
+	// If the object does not exist, there would be no oldRequests either.
+	fuzzerObj := serv.mgr.getFuzzer()
```

Comment:
Can't fuzzerObj be nil here, if the instance has called Connect but did not call Check?

Reply:
If it did not call Check, `runner.requests` would be empty, so there would be no calls to `fuzzerObj`.